### ISSUE TITLE
.IFREF + .PROC in one directive

### ISSUE THREAD

**Issue by vrubleg** [cc65/cc65#1056]
I use combination of .IFREF + .PROC quite often. It looks like this:
```
.IFREF snd_reset
.PROC snd_reset

	LDA #0
	STA APU_CHANCTRL
	RTS

.ENDPROC
.ENDIF
```
Not very convenient and requires to repeat function name twice. It would be much better if there was something like `.IFREFPROC` or even `.WEAKPROC` which would allow to rewrite the code this way:
```
.WEAKPROC snd_reset

	LDA #0
	STA APU_CHANCTRL
	RTS

.ENDPROC
```
The main idea is that this function is included into binary iff it is used in current translation unit.

**Comment by vrubleg** [comment]
Sometimes it is also useful for tables of data also. Now it looks like this:
```
.IFREF nsf_sys_types_tbl
nsf_sys_types_tbl:
	.BYTE NTSC_COUNTDOWN_START, SYS_TYPE_PAL, SYS_TYPE_DENDY
.ENDIF
```
It could look like this:
```
.WEAKDATA nsf_sys_types_tbl
	.BYTE NTSC_COUNTDOWN_START, SYS_TYPE_PAL, SYS_TYPE_DENDY
.ENDDATA
```
So, `.WEAKPROC` and `.WEAKDATA` could be synonyms. If it will make implementation simpler, both of them could be closed by `.ENDWEAK` instead of `.ENDPROC` and `.ENDDATA`.

**Comment by vrubleg** [comment]
I use ca65 for my hobby project (for historical reasons: I started from cc65, and then switched to pure assembly). It is nice, and has quite powerful macros, but there are a few things which chould make using ca65 even more enjoyable, all of them are related to each other: #796, #1056, #1057. At this moment, I'm in the middle of other project, and can't switch to learning internals of ca65 and implementation of these enhancements. But if somebody knows internals of the assembler, likes these ideas, and wants to implement them, I would pay 50$ for each of the tickets (or 200$ if of them are implemented by the same person). Not much, but it's just a "thank you" for making ca65 better.

**Comment by laubzega** [comment]
Hey @vrubleg , please have a look at the PR above. It's not extensively tested, but should be relatively robust against obvious misuse.

Minimal example:
```
.export _main
_main:
;    jmp test
    jmp _main

.WEAKPROC test
    rts
.ENDWEAK
```

I don't see much sense in polluting the command namespace with identically working .WEAKDATA, but if @oliverschmidt or @greg-king5 don't think it's a problem, I can add it, too.

**Comment by greg-king5** [comment]
I don't like that use of the "weak" prefix.  In my mind, "weak" means "create a symbol if it doesn't exist already".  We need a better way to imply "create a name if it's needed".  (But, I can't think of anything.)

**Comment by mrdudz** [comment]
i know "weak symbol" in the exact same meaning as the first post implies. probably from LD

**Comment by Marrin** [comment]
Linker configs are using „weak“ for the type in the SYMBOLS section with the meaning @greg-king5 mentions.

**Comment by vrubleg** [comment]
Oh, great news! This feature must be very convenient for people who use assembly only. I'll test it a bit later.

@greg-king5 Maybe `.LAZYPROC`, then? The term "lazy" is frequently used when something is calculated automatically when the result is actually used. In this case it is somehow similar: the procedure is included when it is actually used (referenced).

**Comment by laubzega** [comment]
Or '.LAZYCODE'? Sounds informal, but covers both proc and data cases.

**Comment by oliverschmidt** [comment]
So far `.IFREFPROC` seems the most natural name to me.

**Comment by vrubleg** [comment]
@oliverschmidt yeah, but `.ENDIFREFPROC` seems like too long. But it is better than combination of `.ENDPROC`+`.ENDIF` in two lines anyway =) Or... Can it be closed by just `.ENDPROC` somehow then, what do you think? It will look like `.IFREFPROC` is just a type of `.PROC`, like there are many different types of `.IF` which are closed by just `.ENDIF`. But it might be harder to implement though. If you don't like the idea, I'm OK with just `.IFREFPROC` and `.ENDIFREFPROC`.

Related to the data counterpart to the `.PROC`, I think I'll just make defines to map `.DATA` to `.PROC` and `.IFREFDATA` to `.IFREFPROC`, so the code will look more semantically correct related to what is actually stored inside of block. Hope, that `.DEFINE` allows to do this, didn't check it yet =)

**Comment by laubzega** [comment]
@vrubleg, beyond naming, any other issues with the patch? Did you have a chance to try it?

**Comment by vrubleg** [comment]
@laubzega I'm sorry, I'm visiting my parents in other country at the moment, with limited access to PC. I'll come back home in a week.

**Comment by vrubleg** [comment]
So, we have two situations, when we want to have a scoped block (`.IFREF` plus `.PROC` or `.SCOPE`), e.g.:
```
.IFREF snd_reset
.PROC snd_reset
	RTS
.ENDPROC
.ENDIF
```
And when we don't want to have a scoped block, e.g.:
```
.IFREF nsf_sys_types_tbl
nsf_sys_types_tbl:
	.BYTE 1, $80, $81
.ENDIF
```

All of these use cases could use such syntax without duplication of labels:
```
.IFREFPROC snd_reset1 ; .IFREF + .PROC
	RTS
.ENDIFREFPROC ; or just .ENDIF?

.IFREFSCOPE snd_reset2 ; .IFREF + .SCOPE
	RTS
.ENDIFREFSCOPE ; or just .ENDIF?

.IFREFLABEL nsf_sys_types_tbl ; .IFREF + usual label without scope
	.BYTE 1, $80, $81
.ENDIFREFLABEL ; or just .ENDIF?
```

Ideas of names:
```
.IFREFPROC   .IFREFSCOPE   .IFREFLABEL   ; literally .IFREF + something.
.SMARTPROC   .SMARTSCOPE   .SMARTLABEL   ; or .SMARTBLOCK instead of .SMARTLABEL, I really like this option.
.OPTPROC     .OPTSCOPE     .OPTLABEL     ; OPT = "optional".
.LAZYPROC    .LAZYSCOPE    .LAZYLABEL    ; Being lazy means that it is included when it is used only.
.WEAKPROC    .WEAKSCOPE    .WEAKLABEL
```

What do you think? It would be really nice to introduce such feature for all the use cases.

**Comment by oliverschmidt** [comment]
Just my two cents:

`.IFREFPROC   .IFREFSCOPE   .IFREFLABEL` - obvious to me - as already stated above.
`.SMARTPROC   .SMARTSCOPE   .SMARTLABEL` - with "smart" I associate more than 'to exist' or 'to not exist'.
`.OPTPROC     .OPTSCOPE     .OPTLABEL` - my first association is for sure 'optimized', not 'optional'.
`.LAZYPROC    .LAZYSCOPE    .LAZYLABEL` - with "lazy" I associate 'later in time', not 'only of necessary'.
`.WEAKPROC    .WEAKSCOPE    .WEAKLABEL` - as already discussed has a different meaning.

**Comment by mrdudz** [comment]
This reminds me of how everyone requests a hotkey for his favourite pet feature in VICE - i dont like the idea at all. It's just creating a mess.

**Comment by vrubleg** [comment]
Additional thoughts. Some way of referencing a label without adding anything to output is required sometimes. Consider this example, when a function has two entry points:
```
.IFREFPROC entry1
	LDA #$10
.ENDIFREFPROC
.IFREFPROC entry2
	; ... some code
	RTS
.ENDIFREFPROC
```
When `entry1` is used, `entry2` should also be included into binary, because execution of the `entry1` proc falls into `entry2`. It could look like this:
```
.IFREFPROC entry1
	LDA #$10
	REF entry2
.ENDIFREFPROC
.IFREFPROC entry2
	; ... some code
	RTS
.ENDIFREFPROC
```
But probably it is possible to do now as a macro.

@mrdudz it is mess when you have to repeat label names twice when you define an optional procedure (and since `.IFREF` exists, it was considered as a legal use case for `ca65`). There might be hundreds of optional functions, and it is always not good when you repeat the same thing a few times without real necessity, and when it actually can be automated.

Some other assemblers include contents of `.PROC` only when it is actually used (e.g. FASM). ca65 has both `.PROC` and `.SCOPE`, and `.PROC` behaves like `.SCOPE`, it is always included. Changing behavior of `.PROC` to be smarter is too much, because it may break compatibility with existing code, but adding new directives should be fine. Anybody who don't like such features can just ignore their existence.

**Comment by kugelfuhr** [comment]
I don't think adding additional identifiers is a good idea, since this is a classic example for use of macros:

```asm
.macro  IFREFPROC name
        .IFREF name
        .PROC name
.endmacro

.macro  ENDIFREFPROC
        .ENDPROC
        .ENDIF
.endmacro

IFREFPROC snd_reset

	LDA #0
	STA APU_CHANCTRL
	RTS

ENDIFREFPROC
```

The assembler will currently say "Conditional assembly branch was never closed" because it checks for open ifs at the end of a macro expansion. My suggestion is to fix this (which is rather easy) so macros can be used instead of adding all sorts of new identifiers that are actually just a combination of existing ones.

**Comment by mrdudz** [comment]
> My suggestion is to fix this (which is rather easy) so macros can be used instead of adding all sorts of new identifiers that are actually just a combination of existing ones.

That sounds like a very good idea, i don't like adding identifiers for more or less random things either

**Comment by kugelfuhr** [comment]
Unfortunately I was a bit too hasty here. It is not as easy as I originally said, since it requires expanding macros while parsing an `.if` branch that is excluded. But I still think it should be doable.
